src/controllers/resource-store/index.tsx

```diff
@@ -104,75 +104,99 @@ export const actions: Actions = {
+      return originalSlice;
     }
 
     dispatch(validateLRUCache(resource, key));
 
-    const pending = {
-      ...slice,
-      data: maxAge === 0 ? null : slice.data,
-      error: maxAge === 0 ? null : slice.error,
+    const promise = getData(
+      { ...routerStoreContext, isPrefetch: !!prefetch },
+      context
+    );
+    const pendingSlice = {
+      ...originalSlice,
+      data: maxAge === 0 ? null : originalSlice.data,
+      error: maxAge === 0 ? null : originalSlice.error,
       loading: true,
-      promise: getData(
-        { ...routerStoreContext, isPrefetch: !!prefetch },
-        context
-      ),
+      promise,
       accessedAt: getAccessedAt(),
     };
 
-    dispatch(setResourceState(type, key, pending));
-
-    const response = {
-      ...pending,
-    };
+    dispatch(setResourceState(type, key, pendingSlice));
 
+    // do not introduce additional awaited function call here, it will cause additional microtasks
+    let resolvedSlice;
     try {
-      response.error = null;
-
       if (timeout) {
         const timeoutGuard = generateTimeGuard(timeout);
-        const maybeData = await Promise.race([
-          pending.promise,
-          timeoutGuard.promise,
-        ]);
+        const maybeData = await Promise.race([promise, timeoutGuard.promise]);
 
         if (!timeoutGuard.isPending) {
-          response.data = null;
-          response.error = new TimeoutError(type);
-          response.loading = true;
-          response.promise = null;
+          // timeout state intentionally looks like hung loading, with exception of the Error
+          // however promise must be null to avoid actually hanging
+          resolvedSlice = {
+            data: null,
+            error: new TimeoutError(type),
+            loading: true,
+            promise: null,
+          };
         } else {
           timeoutGuard.timerId && clearTimeout(timeoutGuard.timerId);
-          response.data = maybeData;
-          response.loading = false;
+          resolvedSlice = {
+            data: maybeData,
+            error: null, // previous error is removed
+            loading: false,
+            promise,
+          };
         }
       } else {
-        response.data = await pending.promise;
-        response.loading = false;
+        resolvedSlice = {
+          data: await promise,
+          error: null, // previous error is removed
+          loading: false,
+          promise,
+        };
       }
-    } catch (e) {
-      response.error = e;
-      response.loading = false;
+    } catch (error) {
+      resolvedSlice = {
+        // previous data is retained
+        promise,
+        error,
+        loading: false,
+      };
     }
 
-    response.expiresAt = getExpiresAt(
-      options.prefetch && maxAge < PREFETCH_MAX_AGE ? PREFETCH_MAX_AGE : maxAge
-    );
+    // do not overwrite a slice that was updated during async
+    const laterSlice = getSliceForResource(getState(), { type, key });
```

Comment:
Are we sure about this? If I'm implementing some sort of optimistic updates, this will discard server data for client-side data that I might have manually set, and which might be incorrect (especially if server returns an error).

Reply:
IMO any optimistic update needs to be done before refresh otherwise it implicitly relies on the async nature of refresh to avoid race condition.

Fundamentally if the `promise` has changed during the async we are in a difference state.

The options as I see it are
1. explicit updates do no effect the `promise`, and the `promise` mismatches `data` (current state before PR)
2. explicit updates set the `promise` consistent with `data`, but preceding async refresh will set it back when it completes
3. explicit updates set the `promise` consistent with `data`, cancels in-flight async refresh

Fixing (1) is the point of this PR so I'd say its a choice between (2) and (3).
Choosing (2) puts us in a position where the `promise` no longer reflects the future value. So is also inconsistent in its own way.
I would specifically chose (3) to force developers to consider the order. It is the only solution that makes `promise` consistent with current `data` and future async `data`.

Reply:
> IMO any optimistic update needs to be done before refresh otherwise it implicitly relies on the async nature of refresh to avoid race condition.

Agree. That is why I would make the async flow win anyway.

> Fundamentally if the promise has changed during the async we are in a difference state.

I'm confused on when this might happen. If we exclude a manual update, then there should be no way for something else to change that promise that is not this async flow itself. 
If this is a way to mitigate #104, then I think we need to fix that bug and not account for that eventuality at all.